### PR TITLE
Introduce libtorch to setup.py build

### PR DESCRIPTION
**Pull request by anderspapitto**
Prior to this diff, there have been two ways of compiling the bulk of the torch codebase. There was no interaction between them - you had to pick one or the other.

1) with setup.py. This method
- used the setuptools C extension functionality
- worked on all platforms
- did not build test_jit/test_api binaries
- did not include the C++ api
- always included python functionality
- produced _C.so

2) with cpp_build. This method
- used CMake
- did not support Windows or ROCM
- was capable of building the test binaries
- included the C++ api
- did not build the python functionality
- produced libtorch.so

This diff combines the two.

1) cpp_build/CMakeLists.txt has become torch/CMakeLists.txt. This build
- is CMake-based
- works on all platforms
- builds the test binaries
- includes the C++ api
- does not include the python functionality
- produces libtorch.so

2) the setup.py build
- compiles the python functionality
- calls into the CMake build to build libtorch.so
- produces _C.so, which has a dependency on libtorch.so

In terms of code changes, this mostly means extending the cmake build to support the full variety of environments and platforms. There are also a small number of changes related to the fact that there are now two shared objects - in particular, windows requires annotating some symbols with dllimport/dllexport, and doesn't allow exposing thread_local globals directly.

### CODE REVIEW

File: torch/CMakeLists.txt [hunk 1]
```diff
@@ -188,60 +86,62 @@ add_custom_command(
   WORKING_DIRECTORY "${TORCH_SRC_DIR}/..")
 
 set(TORCH_SRCS
-  ${TORCH_SRC_DIR}/csrc/autograd/aten_variable_hooks.cpp
-  ${TORCH_SRC_DIR}/csrc/autograd/generated/VariableType.cpp
-  ${TORCH_SRC_DIR}/csrc/autograd/generated/Functions.cpp
-  ${TORCH_SRC_DIR}/csrc/autograd/profiler.cpp
-  ${TORCH_SRC_DIR}/csrc/autograd/saved_variable.cpp
-  ${TORCH_SRC_DIR}/csrc/autograd/grad_mode.cpp
+  ${TORCH_SRC_DIR}/csrc/assertions.cpp
```
Comment (anderspapitto) [inline torch/CMakeLists.txt]: this list has been `M-x sort-lines`ed, no other changes

File: setup.py [hunk 2]
```diff
@@ -711,117 +713,64 @@ def run(self):
 main_libraries = ['shm']
 main_link_args = CAFFE2_LIBS + [NANOPB_STATIC_LIB, PROTOBUF_STATIC_LIB]
 main_sources = [
-    "torch/csrc/PtrWrapper.cpp",
-    "torch/csrc/Module.cpp",
-    "torch/csrc/Generator.cpp",
-    "torch/csrc/Size.cpp",
-    "torch/csrc/Dtype.cpp",
+    "torch/csrc/DataLoader.cpp",
```
Comment (anderspapitto) [inline setup.py]: this list has been `M-x sort-lines`ed, and then all files which are in the TORCH_SRCS list below in the cmake file have been removed

File: .jenkins/pytorch/build.sh [hunk 3]
```diff
@@ -100,5 +100,5 @@ if [[ "$BUILD_TEST_LIBTORCH" == "1" ]]; then
   echo "Building libtorch"
   # NB: Install outside of source directory (at the same level as the root
   # pytorch folder) so that it doesn't get cleaned away prior to docker push.
-  WERROR=1 VERBOSE=1 tools/cpp_build/build_all.sh "$PWD/../cpp-build"
+  TORCH_WERROR=1 VERBOSE=1 tools/cpp_build/build_all.sh "$PWD/../cpp-build"
```
Comment (ezyang) [inline .jenkins/pytorch/build.sh]: What's going on here?

File: aten/CMakeLists.txt [hunk 4]
```diff
@@ -80,14 +80,20 @@ add_subdirectory(src/TH)
 set(TH_CPU_INCLUDE
   # dense
   ${CMAKE_CURRENT_SOURCE_DIR}/src/TH
-  ${CMAKE_CURRENT_SOURCE_DIR}/src/THC
   ${CMAKE_CURRENT_BINARY_DIR}/src/TH
-  ${CMAKE_CURRENT_BINARY_DIR}/src/THC
```
Comment (ezyang) [inline aten/CMakeLists.txt]: Thanks!

File: torch/CMakeLists.txt (hunk 1 above)
Comment (ezyang) [reply torch/CMakeLists.txt]: Thank you, comments like this are much appreciated.

File: torch/CMakeLists.txt [hunk 5]
```diff
@@ -288,47 +188,60 @@ target_compile_options(torch
   -Wno-missing-braces # This warning is buggy
   -Wno-unknown-pragmas)
 
-if ($ENV{WERROR})
+if ($ENV{TORCH_WERROR})
```
Comment (ezyang) [inline torch/CMakeLists.txt]: If this env var got renamed, doesn't the CI env var have to track the change? (Why rename it at all?)

File: torch/CMakeLists.txt [hunk 6]
```diff
@@ -288,47 +188,60 @@ target_compile_options(torch
...
+endif()
+
+set(TH_CPU_INCLUDE
+  # dense
+  ${TORCH_SRC_DIR}/../aten/src/TH
+  ${CMAKE_CURRENT_BINARY_DIR}/../aten/src/TH
+  ${TORCH_SRC_DIR}/../aten/src
+  ${CMAKE_CURRENT_BINARY_DIR}/../aten/src
+  ${CMAKE_BINARY_DIR}/aten/src)
+target_include_directories(torch PRIVATE ${TH_CPU_INCLUDE})
```
Comment (ezyang) [inline torch/CMakeLists.txt]: If you're using the official `caffe2_gpu_library` target, aren't these include paths supposed to be automatically added?

**Review by ezyang — APPROVED** [review]

File: .jenkins/pytorch/build.sh (hunk 3 above)
Comment (anderspapitto) [reply .jenkins/pytorch/build.sh]: now that torch is built via add_subdirectory from caffe2, it's exposed to a new set of `-W...` flags that breaks the build in CI. Given that we already have a tech debt item to clean up and homogenize Werror usage once all the dust settles, having torch not share the exact same WERROR flag with caffe2 is an easy way to unblock this diff.

File: torch/CMakeLists.txt (hunk 5 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: I did change it in CI around usages of cpp_build (explanation is by my comment above)

File: torch/CMakeLists.txt (hunk 6 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: very possibly. I'll try it out

Comment (anderspapitto) [reply torch/CMakeLists.txt]: tried it, didn't work. it may be an opportunity for a future refactor

**Comment by yf225** [comment]
In the previous CI build, Windows exit code was not propagating correctly and the failed build was marked as success (https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-build/10441/console). We need to check that this doesn't happen before merging.

cc. @peterjc123

File: .jenkins/pytorch/build.sh (hunk 3 above)
Comment (goldsborough) [reply .jenkins/pytorch/build.sh]: I think it would be easier/better to use the same set of warning flags for the whole project, and then add the warnings back. Since this build has always been separate, I could do whatever I wanted, including adding all the warnings I wanted. If we're unifying, we should use the same warnings with the same `WERROR` everywhere. So you should instead just remove the additional warnings that break the CI

File: torch/CMakeLists.txt (hunk 5 above)
Comment (goldsborough) [reply torch/CMakeLists.txt]: This should be undone, see comment above.

File: torch/CMakeLists.txt [hunk 7]
```diff
@@ -288,47 +188,60 @@ target_compile_options(torch
   -Wno-missing-braces # This warning is buggy
   -Wno-unknown-pragmas)
 
-if ($ENV{WERROR})
+if ($ENV{TORCH_WERROR})
   target_compile_options(torch PRIVATE -Werror)
+else()
+  target_compile_options(torch PRIVATE -Wno-error)
```
Comment (goldsborough) [inline torch/CMakeLists.txt]: What is this for?

File: torch/CMakeLists.txt [hunk 8]
```diff
@@ -344,45 +257,36 @@ install(TARGETS torch
...
-      # Temporary until ATen tests are built with Caffe2
-      ${TORCH_API_TEST_DIR}/tensor_options.cpp
-      ${TORCH_API_TEST_DIR}/tensor_options_cuda.cpp
-    )
+  target_link_libraries(test_api torch)
 
-      target_include_directories(test_api
-        PUBLIC
-        "${TORCH_SRC_DIR}/../third_party/catch/single_include")
+  target_compile_options(test_api PRIVATE -Wno-error)
```
Comment (goldsborough) [inline torch/CMakeLists.txt]: uh, no?

**Review by goldsborough — DISMISSED** [review]

**Comment by peterjc123** [comment]
@yf225 Yes, we should add them back. As for the build failure, we need to decorate these functions with `torch_API` and make sure `torch_API_exports` is defined.

**Comment by anderspapitto** [comment]
I think most of what's left at this point is adding the CUDA files when USE_CUDA is on

**Comment by ezyang** [comment]
build failures look legit

**Comment by anderspapitto** [comment]
This diff is ready for another round of review. I highlight two points of potential concern.

1) The ROCM builds consistently fail due to 

- timeouts (example https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang3.8-rocm1.7.1-ubuntu16.04-test/2182/console) 
or 
- OOM  (example https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang3.8-rocmnightly-ubuntu16.04-build/2775/console). 

This is particularly annoying because I did fix legit ROCM issues along the way, so I would like to get a clear signal that everything is now correct. However, my initial feeling is that fixing these infra-level resource issues is outside the scope of this diff.

2) On windows, a shared library cannot expose global variables with dllexport. Straightforward application of the new TORCH_API macro results in errors messages like

```
C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\torch/csrc/autograd/profiler.h(172): error C2492: 'thread_id': data with thread storage duration may not have dll interface
```

seen here https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-build/12062/console.

In this diff, I've addressed this issue by moving all the relevant thread_local global variables out of the header files, and instead exposing accessor methods across the shared-object boundary. I'm concerned that this may be a perf problem, if function call overhead is too high a cost to pay to access these profiling-related variables. So, is this a problem, and if so is there a suggested alternative approach?

@ezyang @goldsborough as existing reviewers (also please bring in anyone else with relevant knowledge. E.g. I'm not sure who is most familiar with the profiling code.)

File: tools/autograd/templates/VariableType.h [hunk 9]
```diff
@@ -4,6 +4,8 @@
 
 #include <ATen/ATen.h>
 
+#include "torch/csrc/WindowsTorchApiMacro.h"
```
Comment (goldsborough) [inline tools/autograd/templates/VariableType.h]: nit: prefer <> in the future

File: torch/csrc/autograd/function.cpp [hunk 10]
```diff
@@ -17,7 +17,13 @@
 
 namespace torch { namespace autograd {
 
-thread_local uint64_t Function::next_sequence_nr_ = 0;
+/// Monotonically incrementing (thread local!) counter to supply sequence
+/// numbers.
+thread_local uint64_t Function_next_sequence_nr_ = 0;
```
Comment (goldsborough) [inline torch/csrc/autograd/function.cpp]: Why can't the thread local variables be class-static? Why do they have to be global? Windows?

File: torch/csrc/autograd/profiler.h [hunk 11]
```diff
@@ -162,80 +162,19 @@ enum class ProfilerState {
...
-#ifdef USE_CUDA
-    nvtxRangePop();
-#else
-    throw std::logic_error("popRange called with NVTX tracing, but compiled without CUDA");
-#endif
-  } else {
-    getEventList().record(EventKind::PopRange, std::string(), thread_id, state == ProfilerState::CUDA);
-  }
-}
+RangeEventList& getEventList();
```
Comment (goldsborough) [inline torch/csrc/autograd/profiler.h]: What happened here?

File: torch/csrc/autograd/init.cpp [hunk 12]
```diff
@@ -26,7 +26,9 @@ PyObject * THPAutograd_initExtension(PyObject *_unused)
 
   py::class_<torch::autograd::profiler::Event>(m,"ProfilerEvent")
   .def("kind",&torch::autograd::profiler::Event::kind)
-  .def("name",&torch::autograd::profiler::Event::name)
```
Comment (goldsborough) [inline torch/csrc/autograd/init.cpp]: What happened here?

File: torch/csrc/autograd/grad_mode.h [hunk 13]
```diff
@@ -3,14 +3,8 @@
 namespace torch { namespace autograd {
 
 struct GradMode {
-  static bool is_enabled() {
```
Comment (goldsborough) [inline torch/csrc/autograd/grad_mode.h]: Why was this required?

File: torch/CMakeLists.txt [hunk 14]
```diff
@@ -265,73 +183,145 @@ if (NOT NO_API)
...
+    -Wextra
+    -Wno-unused-parameter
+    -Wno-missing-field-initializers
+    -Wno-write-strings
+#    -Wno-zero-length-array
+    -Wno-unknown-pragmas
+    # This is required for Python 2 declarations that are deprecated
+    # in 3.
+    -Wno-deprecated-declarations
+    # Python 2.6 requires -fno-strict-aliasing see
```
Comment (goldsborough) [inline torch/CMakeLists.txt]: I'm curious if we couldn't have some flag to know whether this build is for Python, or not for Python? If not for Python, we wouldn't need this `no-strict-aliasing` flag for example.

File: torch/CMakeLists.txt [hunk 15]
```diff
@@ -265,73 +183,145 @@ if (NOT NO_API)
...
+    )
+else()
+  target_compile_options(torch PRIVATE
+    -std=c++11
+    -Wall
+    -Wextra
+    -Wno-unused-parameter
+    -Wno-missing-field-initializers
+    -Wno-write-strings
+#    -Wno-zero-length-array
```
Comment (goldsborough) [inline torch/CMakeLists.txt]: What's with `Wno-zero-length-array`?

File: torch/CMakeLists.txt [hunk 16]
```diff
@@ -347,45 +337,76 @@ install(TARGETS torch
...
+  target_include_directories(test_api
+    PUBLIC
+    "${TORCH_SRC_DIR}/../third_party/catch/single_include"
+    ${ATen_CPU_INCLUDE})
+
+  target_link_libraries(test_api torch ${TORCH_CUDA_LIBRARIES})
+
+  if (APPLE)
+    target_compile_options(test_api PRIVATE
+      -Wno-error=unknown-warning-option
```
Comment (goldsborough) [inline torch/CMakeLists.txt]: I'm actually not a fan of this. I want my build to be as clean as possible. If we're not going to fix the warning, let's disable it and not have it spew in our build (i.e. I'd rather `Wno-missing-braces`). I or someone else will fix the warnings eventually.

That said, `Wno-reorder` is just plain evil. It's just inviting free bugs when you use a member in the initialization of another member.
`Wmaybe-uninitialized` should be fixed with `(void)variable` and we have in the past.

Why are you adding these new warning disablements anyway? The code being compiled didn't change, did it?

**Review by goldsborough — COMMENTED** [review]
Curious about some things, especially the newly disabled warnings

File: torch/csrc/autograd/function.cpp (hunk 10 above)
Comment (anderspapitto) [reply torch/csrc/autograd/function.cpp]: Yes, windows. See my point 2 in my big comment above in the main thread.

Specifically the restriction here isn't about class-static vs global, but rather that it can't be in the header (and this is a smaller change than moving the entire class out of the header)

File: torch/csrc/autograd/profiler.h (hunk 11 above)
Comment (anderspapitto) [reply torch/csrc/autograd/profiler.h]: for all of these, see my top-level comment about windows

File: torch/CMakeLists.txt (hunk 14 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: Regarding all of the various warnings flags - I haven't introduced any code constructs that elicit new warnings and/or errors. They're all needed by pre-existing code, and the Werror violations are exposed by the fact that all compliations/build environments are now driven through this unified codepath.

I claim that it's slightly out-of-scope of this diff to fix all the underlying, pre-existing errors, beyond what's needed to get a green CI. I'd rather record this in the unification-code-debt list and churn through them in a follow-up diff.

File: torch/CMakeLists.txt (hunk 15 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: It breaks one of the build configurations, I think maybe it was that clang doesn't acknowledge that this flag exists?

File: torch/CMakeLists.txt (hunk 16 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: agreed, however see my response above

File: torch/CMakeLists.txt (hunk 15 above)
Comment (goldsborough) [reply torch/CMakeLists.txt]: It shouldn't be commented out either way I think, let's remove it if we don't need it

File: torch/CMakeLists.txt (hunk 14 above)
Comment (goldsborough) [reply torch/CMakeLists.txt]: What are the new "compilations"/"build environments"? Just to understand

File: torch/CMakeLists.txt (hunk 16 above)
Comment (goldsborough) [reply torch/CMakeLists.txt]: You don't have to fix the warnings, but do please change them to `-Wno-unknown-warning-option` etc. instead of `-Wno-error=...` so that they don't show up in the build

File: tools/autograd/templates/VariableType.h [hunk 17]
```diff
@@ -49,7 +51,7 @@ struct VariableType final : public at::Type {
   virtual std::unique_ptr<at::Storage> unsafeStorageFromTH(void * th_pointer, bool retain) const override;
   virtual at::Tensor unsafeTensorFromTH(void * th_pointer, bool retain) const override;
 
-  static at::Type* getType(const at::Type& baseType);
+  TORCH_API static at::Type* getType(const at::Type& baseType);
```
Comment (ezyang) [inline tools/autograd/templates/VariableType.h]: Maybe we should just `TORCH_API` the entire class?

File: tools/build_pytorch_libs.bat [hunk 18]
```diff
@@ -175,6 +175,9 @@ goto:eof
   cmake .. %CMAKE_GENERATOR_COMMAND% ^
                   -DCMAKE_BUILD_TYPE=%BUILD_TYPE% ^
                   -DBUILD_CAFFE2=OFF ^
+                  -DBUILD_TORCH="%BUILD_TORCH%" ^
+                  -DNVTOOLEXT_HOME="%NVTOOLEXT_HOME%" ^
+                  -DNO_API=ON ^
```
Comment (ezyang) [inline tools/build_pytorch_libs.bat]: @goldsborough Is this intentional?

File: torch/CMakeLists.txt [hunk 19]
```diff
@@ -265,73 +183,145 @@ if (NOT NO_API)
...
+    -Wno-missing-braces
+    )
+endif()
+
+if (MSVC)
+elseif ($ENV{WERROR})
   target_compile_options(torch PRIVATE -Werror)
 endif()
 
+if (MSVC)
```
Comment (ezyang) [inline torch/CMakeLists.txt]: I don't understand: why is this MSVC only?

File: torch/csrc/autograd/function.cpp (hunk 10 above)
Comment (ezyang) [reply torch/csrc/autograd/function.cpp]: Also, to be clear, it is not *static* variables that cause the problem, it is *thread-local* variables. (And if you API a class, it implicitly APIs all of the static variables in it, which is the proximal problem).

This change is fine, as far as I'm concerned.

**Comment by ezyang** [comment]
Would it be possible to get a detailed break down of the semantic changes that were made in plain English? E.g., something similar to the comment at https://github.com/pytorch/pytorch/pull/9358 ; it will help us reviewers out a lot.

File: torch/CMakeLists.txt [hunk 20]
```diff
@@ -265,73 +183,145 @@ if (NOT NO_API)
...
+  else()
+    set(TORCH_CUDA_LIBRARIES
+      ${CUDA_CUDA_LIB}
+      ${CUDA_NVRTC_LIB}
+      ${CUDA_TOOLKIT_ROOT_DIR}/lib64/libnvToolsExt.so
+      ${CUDA_LIBRARIES})
+  endif()
+
+  if(MSVC OR APPLE)
+    target_link_libraries(torch caffe2_gpu_library ${TORCH_CUDA_LIBRARIES})
```
Comment (ezyang) [inline torch/CMakeLists.txt]: Why is it `TORCH_CUDA_LIBRARIES` on MSVC/Apple but `CUDA_LIBRARIES` on Linux?

File: torch/CMakeLists.txt [hunk 21]
```diff
@@ -347,45 +337,76 @@ install(TARGETS torch
   LIBRARY DESTINATION "${TORCH_INSTALL_LIB_DIR}"
   ARCHIVE DESTINATION "${TORCH_INSTALL_LIB_DIR}")
 
-if (TORCH_BUILD_TEST)
-  # JIT Tests. TODO: Put into test/cpp/jit folder
-
+# JIT Tests. TODO: Put into test/cpp/jit folder
```
Comment (ezyang) [inline torch/CMakeLists.txt]: Don't we still have a flag that toggles if tests build or not? (Or am I misremembering?)

File: torch/CMakeLists.txt [hunk 22]
```diff
@@ -347,45 +337,76 @@ install(TARGETS torch
...
+    target_compile_options(torch PRIVATE -Wno-error=maybe-uninitialized)
+  endif()
+
+  if (APPLE)
+    target_compile_options(torch PRIVATE -Wno-error=unknown-warning-option)
+  endif()
+endif()
 
-    target_link_libraries(test_api torch)
+if (NOT NO_API AND NOT USE_ROCM)
```
Comment (ezyang) [inline torch/CMakeLists.txt]: Why not absorb this if block into the one that's immediately above it?

File: tools/autograd/templates/VariableType.h (hunk 9 above)
Comment (anderspapitto) [reply tools/autograd/templates/VariableType.h]: ok, i'll switch it. what's the guideline to pick one or the other? It seems that quotes are more common in the codebase

```
(venv)[andersj@devgpu201.prn2 ~/local/pytorch] git grep 'include "torch'|wc
    911    1822   69467
(venv)[andersj@devgpu201.prn2 ~/local/pytorch] git grep 'include <torch'|wc
    238     479   16250
```

File: torch/CMakeLists.txt (hunk 16 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: ++

File: torch/CMakeLists.txt (hunk 14 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: for example, the fact that torch is built by cmake on windows - this was never the case before. Or, test_api being compiled with the set of compiler flags taken from setup.py

File: torch/CMakeLists.txt (hunk 15 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: ++

File: aten/src/ATen/ATenGeneral.h [hunk 23]
```diff
@@ -1,7 +1,9 @@
 #pragma once
 
 #ifdef _WIN32
-# if defined(ATen_cpu_EXPORTS) || defined(caffe2_EXPORTS)
+# if defined(torch_EXPORTS)
```
[the remark this answers is not in the document]
Comment (anderspapitto) [reply aten/src/ATen/ATenGeneral.h]: I'll double check this

File: setup.py [hunk 24]
```diff
@@ -673,7 +677,7 @@ def run(self):
 library_dirs.append(lib_path)
 
 # we specify exact lib names to avoid conflict with lua-torch installs
-CAFFE2_LIBS = [os.path.join(lib_path, 'libcaffe2.so')]
+CAFFE2_LIBS = [os.path.join(lib_path, 'libcaffe2.so'), os.path.join(lib_path, 'libtorch.so')]
```
[the remark this answers is not in the document]
Comment (anderspapitto) [reply setup.py]: I'm not quite sure what you mean. Whenever we would have previously produced (_C.so), we now produce (_C.so, libtorch.so)

File: tools/autograd/templates/VariableType.h (hunk 17 above)
Comment (anderspapitto) [reply tools/autograd/templates/VariableType.h]: does that work? I can shoot off a test run and switch to that if it passes

File: tools/build_pytorch_libs.bat (hunk 18 above)
Comment (anderspapitto) [reply tools/build_pytorch_libs.bat]: the API was only ever compiled through cpp_build, not through setup.py, and cpp_build had never worked on windows

File: torch/CMakeLists.txt (hunk 19 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: windows doesn't have implicit transitive shared libraries. The setup.py build does the same, with this comment (which i can copy over)

    # Windows needs direct access to ONNX libraries as well
    # as through Caffe2 library

File: torch/CMakeLists.txt (hunk 20 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: I'm unclear on the exact underlying situation, but linking against the additional libraries breaks the build in some of our CI environments. I don't remember the error, would have to kick of another test run to reproduce it.

note that we do need the extra libraries when building executables (see test_api and test_jit below)

File: torch/CMakeLists.txt (hunk 21 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: it used to default to ON and there were no consumers that set it OFF. I can reintroduce it if you like

File: torch/CMakeLists.txt (hunk 22 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: ++

File: torch/csrc/autograd/function.cpp (hunk 10 above)
Comment (anderspapitto) [reply torch/csrc/autograd/function.cpp]: ++

File: torch/CMakeLists.txt (hunk 15 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: kicked off this test run to check exactly what it breaks https://github.com/pytorch/pytorch/pull/9427

File: aten/src/ATen/ATenGeneral.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (anderspapitto) [reply aten/src/ATen/ATenGeneral.h]: kicked off https://github.com/pytorch/pytorch/pull/9428 to test reverting this

File: tools/autograd/templates/VariableType.h (hunk 17 above)
Comment (anderspapitto) [reply tools/autograd/templates/VariableType.h]: kicked off this test run to check https://github.com/pytorch/pytorch/pull/9429

File: tools/autograd/templates/VariableType.h (hunk 9 above)
Comment (goldsborough) [reply tools/autograd/templates/VariableType.h]: I will make a codemod soon! We just had to change some includes that were not absolute paths (in which case <> is incorrect), but now I'll make that codemod

File: torch/CMakeLists.txt (hunk 15 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: error is at https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-trusty-py3.6-gcc7-build/1833/console

File: torch/CMakeLists.txt (hunk 20 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: it causes this failure https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-cuda8-cudnn6-py3-build/14172/console

which had something to do with libtorch.so not being able to find libcuda.so

File: setup.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply setup.py]: This is `CAFFE2_LIBS`, right? I was under the impression that Caffe2 didn't have a dependency on `libtorch.so` in current master. That's what this line looks like it's doing.

File: tools/build_pytorch_libs.bat (hunk 18 above)
Comment (ezyang) [reply tools/build_pytorch_libs.bat]: OK, but morally, there isn't any reason why C++ API shouldn't work on Windows, except maybe because we're not actively supporting at the moment. Deserves a comment, at least.

File: torch/CMakeLists.txt (hunk 19 above)
Comment (ezyang) [reply torch/CMakeLists.txt]: It's better practice to not rely on implicit transitive shared libraries, even on Linux.

File: torch/CMakeLists.txt (hunk 21 above)
Comment (ezyang) [reply torch/CMakeLists.txt]: If we always build tests, then this is harmless, but I was under the impression that some people liked to do builds without having to build tests (I'm not one of them.)

File: torch/CMakeLists.txt (hunk 20 above)
Comment (ezyang) [reply torch/CMakeLists.txt]: OK... I think this merits further investigation, but it doesn't have to be in this PR. I'd recommend commenting about it.

File: torch/CMakeLists.txt (hunk 19 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: ok, I'll send a test run to see it I can remove the `if (MSVC)`. If it doesn't work though, I claim it's out of scope of this diff to invest more time changing it, since the behavior here is unchanged (and was intentional according to that comment).

File: tools/build_pytorch_libs.bat (hunk 18 above)
Comment (anderspapitto) [reply tools/build_pytorch_libs.bat]: I'll open an issue when I merge

File: torch/CMakeLists.txt (hunk 20 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: ++ I'll add a comment

File: torch/CMakeLists.txt (hunk 21 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: ++ I'll re-add it then

File: setup.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (anderspapitto) [reply setup.py]: CAFFE2_LIBS doesn't have to do with compiling caffe2 itself - the only use of this variable is to be added to the link dependencies of _C.so. However given the variable name I guess I am abusing this a bit, so I'll take libtorch out of this list and just add it to main_link_args directly

File: torch/CMakeLists.txt (hunk 19 above)
Comment (anderspapitto) [reply torch/CMakeLists.txt]: does not work out of the box https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-xenial-py3-clang5-asan-test1/3865/console

**Comment by ebetica** [comment]
@anderspapitto I'm using the PyTorch C++ API in csrc/api, and I used to rely on ./build_all.sh to build the libs before linking (in open source). However, this PR got rid of that, and the command seems quite cumbersome:

```
export BUILD_TYPE=relwithdebinfo; export LIBTORCH_BUILDPATH=build/libtorch; ./build_caffe2.sh; ./build_libtorch.sh
```

Any chance it could be brought back and supported in the tests again so it does not break?
